File: tags/fuwjin-0.9.6/dinah/src/main/java/org/fuwjin/dinah/CachedFunctionProvider.java
package org.fuwjin.dinah;

import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;
import org.fuwjin.dinah.function.AbstractFunction;

/**
 * Buffering function provider to reduce function inspection.
 */
public class CachedFunctionProvider extends AbstractFunctionProvider {
   private final AbstractFunctionProvider[] providers;
   private final Set<String> categories = new HashSet<String>();
   private final Map<String, AbstractFunction> functions = new HashMap<String, AbstractFunction>();

   /**
    * Creates a new instance.
    * @param providers the set of providers
    */
   public CachedFunctionProvider(final AbstractFunctionProvider... providers) {
      this.providers = providers;
   }

   @Override
   public Map<String, AbstractFunction> getFunctions(final String category) {
      if(!categories.contains(category)) {
         for(final AbstractFunctionProvider provider: providers) {
            for(final AbstractFunction func: provider.getFunctions(category).values()) {
               add(functions, func);
            }
         }
         categories.add(category);
      }
      return functions;
   }
}

File: tags/fuwjin-0.9.7.2/tweed/tweed.LD/provider/CachedFunctionProvider.cs
using System;
using System.Collections.Generic;
using Tweed.LD.Function;
namespace Tweed.LD.Provider
{
	public class CachedFunctionProvider: FunctionProvider
	{
		private FunctionProvider[] providers;
		private IDictionary<string, string> categories = new Dictionary<string, string>();
		private IDictionary<string, Function.Function> functions = new Dictionary<string, Function.Function>();

		public CachedFunctionProvider (params FunctionProvider[] providers)
		{
			this.providers = providers;
		}
		
		public override IDictionary<string, Function.Function> GetFunctions (string category)
		{
			if(!categories.ContainsKey(category))
			{
				foreach(FunctionProvider provider in providers){
					foreach(Function.Function func in provider.GetFunctions(category).Values){
						Add(functions, func);
					}
				}
				categories.Add(category, category);
			}
			return functions;
		}
		
		
	}
}

